File: 장세훈/LeetCode/medium/CombinationSum.java
import java.util.*;

class CombinationSum {
    /*
        39.  Combination Sum
        https://leetcode.com/problems/combination-sum/description/
     */


    List<List<Integer>> answer = new ArrayList<>();

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        dfs(candidates, new ArrayList<>(), target, 0, 0);
        return answer;
    }

    void dfs(int[] candidates, List<Integer> list, int target, int sum, int start) {
        if (sum > target) return;
        if (sum == target) {
            answer.add(List.copyOf(list));
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            list.add(candidates[i]);
            dfs(candidates, list, target, sum + candidates[i], i);
            list.removeLast();
        }
    }
}

File: 정수빈/LeetCode/medium/combination-sum.py
# https://leetcode.com/problems/combination-sum/
# 설명: 백트레킹을 이용하여 리스트 요소들의 합 == target 만드는 문제입니다. 
# 시간: 8분 21초
# 난이도: medium

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()

        ans = []
        def bk(total, candids):
            if total == target:
                candids
                ans.append(sorted(candids))
                return 
            elif total > target:
                return 
            
            for i in range(len(candidates)):
                can = candidates[i]
                bk(total + can, candids + [can])
        
        bk(0, [])

        ans = list(set(map(lambda x: tuple(x), ans)))
        return ans